ui/src/utils/async_task_queue.rs

```diff
@@ -0,0 +1,128 @@
+use common::{
+    icons::outline::Shape as Icon,
+    state::pending_message::PendingMessage,
+    warp_runner::{ui_adapter::MessageEvent, WarpEvent},
+    WARP_EVENT_CH,
+};
+use dioxus_core::ScopeState;
+use dioxus_hooks::{use_ref, UseRef};
+use futures::{Future, StreamExt};
+use tokio::sync::{
+    mpsc::{UnboundedReceiver, UnboundedSender},
+    Mutex,
+};
+
+use once_cell::sync::Lazy;
+use std::sync::Arc;
+use uuid::Uuid;
+use warp::raygun::{AttachmentEventStream, AttachmentKind, Location};
+
+pub enum ListenerAction {
+    ToastAction {
+        title: String,
+        content: String,
+        icon: Option<Icon>,
+        timeout: u32,
+    },
+}
+
+pub struct ListenerChannel {
+    pub tx: UnboundedSender<ListenerAction>,
+    pub rx: Arc<Mutex<UnboundedReceiver<ListenerAction>>>,
+}
+
+// Channel for actions that should be done on the main thread
+pub static ACTION_LISTENER: Lazy<ListenerChannel> = Lazy::new(|| {
+    let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+    ListenerChannel {
+        tx,
+        rx: Arc::new(Mutex::new(rx)),
+    }
+});
+
+#[derive(Clone, PartialEq)]
+pub struct AsyncRef<T> {
+    inner_ref: Option<Vec<T>>,
+}
+
+impl<T> AsyncRef<T> {
+    /// Appends a value to this queue
+    pub fn append(&mut self, value: T) {
+        match self.inner_ref.as_mut() {
+            Some(current) => {
+                current.push(value);
+            }
+            None => self.inner_ref = Some(vec![value]),
+        };
+    }
+}
+
+/// Create a handler for an async queue
+/// Everytime a value gets added to the queue the future will be spawned when it rerenders
+pub fn async_queue<T: 'static + Send, Fut>(
+    cx: &ScopeState,
+    fut: impl Fn(T) -> Fut,
+) -> &UseRef<AsyncRef<T>>
+where
+    Fut: Future<Output = ()> + Send + 'static,
+{
+    let queue_ref: &UseRef<AsyncRef<T>> = use_ref(cx, || AsyncRef { inner_ref: None });
+    if let Some(queue) = queue_ref.write_silent().inner_ref.take() {
+        for entry in queue {
+            let future = fut(entry);
+            tokio::spawn(future);
+        }
+    }
+    queue_ref
+}
```

Comment:
I wonder if this should be separated out so we can track the tasks being spawned and be able to cancel them if needed. Also is there any case where it doesnt rerenders? If so, that can pose a problem since that mean the future wouldnt be polled and no progress would be made (pretty much leaving it at a standstill).

Reply:
As I think about it, wouldnt `cx.spawn` or `cx.push_future` work in this case since i believe they track the futures internally?

Reply:
no. both `cx.spawn` and `cx.push_future` spawn a future on the main thread. 
also the futures dont cause a rerender. rather the way it works is by adding things to the `AsyncRef` aka you want to do something off thread (which causes a rerender) you cause `tokio::spawn`

Reply:
Hmm i would have to check but i thought they push the task into tokio on desktop (since the single threaded part of it is for web side of it iirc) and poll it there until its complete

Reply:
ye no. dioxus is completly single threaded

Reply:
Hmm I guess ill look more into how dioxus is polling futures